### DISCUSSION TITLE
Duplicated edges in data object and hard constraints

### DISCUSSION THREAD

**Discussion by 101AlexMartin** [pyg-team/pytorch_geometric#9236] [Q&A]
Hello,

I'm dealing with a problem in which the target is to predict the massflow along the edges connecting the nodes. In order to create the graph, I use `networkx `to later map a _networkx.classes.graph.Graph_ to a _torch_geometric.data.data.Data_, by making use of `from_network`. I have a couple of questions related to this step:

1. I check that the created pytorch geometric object duplicates the edges. For instance, in the original networkx graph there is a single edge between nodes 0 and 1, i.e. (0,1). However, after the application of `from_network`, there are two edges among these two nodes, i.e. (0,1) and (1,0). Why is this? (I assume it is because each node is treated as a datapoint when training a GNN(?)).
2. When it comes to the training phase, the fact that edges are duplicated can impose a physics violation. Since every node is evaluated independently, the GNN does not impose that the predicted value for edge (0,1) should be the same as for the edge (1,0), but of opposite sign. Is there any way to impose this as a hard constraint?

**Comment by rusty1s** [comment]
1. Yes, PyG treats an undirected edge as a bidirectional edge, meaning that there will be message passing flow from source to destination node and vice versa.
2. That's correct (and expected) since the features of the node itself and its neighbors might be completely different. IMO, there is no way to hard constraint this since it is a property inherent to GNNs (for a strongly connected graph, this would mean that all output representations across all nodes would be equivalent).
